session.go

```diff
@@ -95,52 +114,164 @@ func (s *session) Serve() error {
 		return err
 	}
 
+	s.stmts = map[string]*nodes.PrepareStmt{}
+	s.pendingStmts = map[string]*nodes.PrepareStmt{}
+	s.portals = map[string]*portal{}
 	t := protocol.NewTransport(s.Conn)
 
 	// query-cycle
 	for {
-		msg, err := t.NextFrontendMessage()
+		msg, ts, err := t.NextFrontendMessage()
 		if err != nil {
 			return err
 		}
 
-		switch v := msg.(type) {
-		case *pgproto3.Terminate:
-			s.Conn.Close()
-			return nil // client terminated intentionally
-		case *pgproto3.Query:
-			q := &query{
-				transport: t,
-				sql:       v.String,
-				queryer:   s.Server,
-				execer:    s.Server,
-			}
-			err = q.Run(s)
-			if err != nil {
-				return err
-			}
-		case *pgproto3.Describe:
-			err = t.Write(protocol.ErrorResponse(fmt.Errorf("not implemented")))
-			if err != nil {
-				return err
-			}
-		case *pgproto3.Parse:
-			err = t.Write(protocol.ErrorResponse(fmt.Errorf("not implemented")))
-			if err != nil {
-				return err
-			}
-		case *pgproto3.Bind:
-			err = t.Write(protocol.ErrorResponse(fmt.Errorf("not implemented")))
-			if err != nil {
-				return err
+		s.handleTransactionState(ts)
+		err = s.handleFrontendMessage(t, msg)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (s *session) handleFrontendMessage(t *protocol.Transport, msg pgproto3.FrontendMessage) (err error) {
+	var res []protocol.Message
+	switch v := msg.(type) {
+	case *pgproto3.Terminate:
+		s.Conn.Close()
+		return nil // client terminated intentionally
+	case *pgproto3.Query:
+		q := &query{
+			transport: t,
+			sql:       v.String,
+			queryer:   s.Server,
+			execer:    s.Server,
+		}
+		err = q.Run(s)
+		if err != nil {
+			res = append(res, protocol.ParseComplete)
+		}
+	case *pgproto3.Describe:
+		res, err = s.describe(msg.(*pgproto3.Describe))
```

Comment:
You can reuse `v` here and in the following lines, like you did in `case *pgproto3.Query`.
```suggestion
		res, err = s.describe(v)
```